experimental/collator/src/provider.rs

```diff
@@ -0,0 +1,266 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+// The reordering algorithms in this file are adapted from ICU4C and,
+// therefore, are subject to the ICU license as described in LICENSE.
+
+use crate::data_ce_to_primary;
+use crate::CollationElement;
+use crate::CollationElement32;
+use crate::Tag;
+use icu_char16trie::char16trie::Char16TrieIterator;
+use icu_codepointtrie::CodePointTrie;
+use icu_provider::{yoke, zerofrom};
+use zerovec::ule::AsULE;
+use zerovec::ule::RawBytesULE;
+use zerovec::ZeroVec;
+
+#[cfg(feature = "provider_serde")]
+use serde;
+
+use super::CaseFirst;
+use super::MaxVariable;
+
+pub mod key {
+    use icu_provider::{resource_key, ResourceKey};
+    pub const COLLATION_DATA_KEY: ResourceKey = resource_key!("collator/data@1");
+    pub const COLLATION_DIACRITICS_KEY: ResourceKey = resource_key!("collator/dia@1");
+    pub const COLLATION_JAMO_KEY: ResourceKey = resource_key!("collator/jamo@1");
+    pub const COLLATION_METADATA_KEY: ResourceKey = resource_key!("collator/meta@1");
+    pub const COLLATION_REORDERING_KEY: ResourceKey = resource_key!("collator/reord@1");
+
+    pub const ALL_KEYS: [ResourceKey; 5] = [
+        COLLATION_DATA_KEY,
+        COLLATION_DIACRITICS_KEY,
+        COLLATION_JAMO_KEY,
+        COLLATION_METADATA_KEY,
+        COLLATION_REORDERING_KEY,
+    ];
+}
+
+#[icu_provider::data_struct(CollationDataV1Marker = "collator/data@1")]
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Serialize, serde::Deserialize)
+)]
```

Comment:
Also for your other structs:
```suggestion
#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
```

Reply:
Done.